cosi/cosi.go

```diff
@@ -1,450 +1,399 @@
-// have to set the mask using `SetMask` method. By default, all participants are
-// designated as participating. If you wish to specify which co-signers are
-// participating, use NewCosiWithMask
-func NewCosi(suite abstract.Suite, private abstract.Scalar, publics []abstract.Point) *CoSi {
-	cosi := &CoSi{
-		suite:   suite,
-		private: private,
-	}
-	// Start with an all-disabled participation mask, then set it correctly
-	cosi.mask = newMask(suite, publics)
-	return cosi
-}
-
-// CreateCommitment creates the commitment of a random secret generated from the
-// given s stream. It returns the message to pass up in the tree. This is
-// typically called by the leaves.
-func (c *CoSi) CreateCommitment(s cipher.Stream) abstract.Point {
-	c.genCommit(s)
-	return c.commitment
+// Commit returns a random scalar v, generated from the given cipher stream, and
+// a corresponding commitment V = [v]G.
+func Commit(suite abstract.Suite, s cipher.Stream) (abstract.Scalar, abstract.Point) {
+	var stream = s
+	if s == nil {
+		stream = random.Stream
+	}
+	random := suite.Scalar().Pick(stream)
+	commitment := suite.Point().Mul(nil, random)
+	return random, commitment
 }
 
-// Commit creates the commitment / secret as in CreateCommitment and it also
-// aggregate children commitments from the children's messages.
-func (c *CoSi) Commit(s cipher.Stream, subComms []abstract.Point) abstract.Point {
-	// generate our own commit
-	c.genCommit(s)
-
-	// add our own commitment to the aggregate commitment
-	c.aggregateCommitment = c.suite.Point().Add(c.suite.Point().Null(), c.commitment)
-	// take the children commitments
-	for _, com := range subComms {
-		c.aggregateCommitment.Add(c.aggregateCommitment, com)
+// AggregateCommitments returns the sum of the given commitments and the
+// bitwise OR of the corresponding masks.
+func AggregateCommitments(suite abstract.Suite, commitments []abstract.Point, masks [][]byte) (abstract.Point, []byte, error) {
+	if len(commitments) != len(masks) {
+		return nil, nil, errors.New("mismatching lengths of commitment and mask slices")
 	}
-	return c.aggregateCommitment
-
+	aggCom := suite.Point().Null()
+	aggMask := make([]byte, len(masks[0]))
+	var err error
+	for i := 0; i < len(commitments); i++ {
+		aggCom = suite.Point().Add(aggCom, commitments[i])
+		aggMask, err = AggregateMasks(aggMask, masks[i])
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	return aggCom, aggMask, nil
 }
 
-// CreateChallenge creates the challenge out of the message it has been given.
-// This is typically called by Root.
-func (c *CoSi) CreateChallenge(msg []byte) (abstract.Scalar, error) {
-	// H( Commit || AggPublic || M)
+// Challenge creates the collective challenge from the given aggregate
+// commitment V, aggregate public key A, mask Z, and message M, i.e., it
+// returns c = H(V || A || Z || M).
+func Challenge(suite abstract.Suite, commitment abstract.Point, mask *Mask, message []byte) (abstract.Scalar, error) {
```

Comment:
Because I need to hash the aggregate public key `A` which is stored in `Mask`. For the aggregation of masks in `AggregateCommitment` it is not (yet) relevant what the (intermediate) aggregate public key is since the leader sends the final aggregated mask down to all nodes so that they can compute the final aggregate public key for the challenge via `SetMask()` on the `Mask` struct that every node has initialized at the beginning of the protocol. 

Alternatively, we could (1) compute the aggregate public key outside of `Challenge` using the aggregate mask or (2) inside of `Challenge` which would require to give the list of all public keys and the final aggregated mask to `Challenge`.

Don't know which of all those variants is better. Thoughts?